diff -ru libtar-1.2.11/lib/decode.c libtar-1.2.11-new/lib/decode.c
--- libtar-1.2.11/lib/decode.c	2003-01-07 02:40:59.000000000 +0100
+++ libtar-1.2.11-new/lib/decode.c	2003-10-03 15:02:44.000000000 +0200
@@ -26,7 +26,7 @@
 char *
 th_get_pathname(TAR *t)
 {
-	char filename[MAXPATHLEN];
+	static char filename[MAXPATHLEN];
 
 	if (t->th_buf.gnu_longname)
 		return t->th_buf.gnu_longname;
@@ -35,11 +35,11 @@
 	{
 		snprintf(filename, sizeof(filename), "%.155s/%.100s",
 			 t->th_buf.prefix, t->th_buf.name);
-		return strdup(filename);
+		return filename;
 	}
 
 	snprintf(filename, sizeof(filename), "%.100s", t->th_buf.name);
-	return strdup(filename);
+	return filename;
 }
 
 
